### DISCUSSION TITLE
I want to use a Java Agent to collect metrics, manually implement custom metrics, and have Prometheus scrape them.

### DISCUSSION THREAD

**Discussion by shinozakisatoshi11** [open-telemetry/opentelemetry-java-instrumentation#12801] [Q&A]
I am Japanese, so please excuse my limited English.

As mentioned in the title, I want to achieve this. However, since Prometheus uses a pull model, it resulted in having two servers. I've tried various approaches to avoid this, but I was unsuccessful. I apologize for my limited knowledge of Java agents, but is there a way to use this configuration or something similar within the application?

Things I have tried:
Attempted to obtain the Java agent provider (exporter).

In javaagent, there are some traces that can be annotated (@withspan) to affect the agent, but no metrics?

**Comment by laurit** [comment]
when using agent use `GlobalOpenTelemetry.get()` to obtain the `OpenTelemetry` instance configured by the agent. When you create your metrics instruments through that instance the metrics will be exported by the agent. If you desire to use prometheus configure the agent to use prometheus exporter for metrics. See https://opentelemetry.io/docs/languages/java/configuration/#properties-exporters set `otel.metrics.exporter` to `prometheus`.

> In javaagent, there are some traces that can be annotated (@withspan) to affect the agent, but no metrics?

currently we don't provide such annotations for recording metrics

**Reply by shinozakisatoshi11** [reply]
Thank you for your response.

It may seem absurd, but the current implementation is as follows. It works without issues, and metrics are being sent to Prometheus. However, I don't want to set up two servers. Is there a way around this?

・Configuration for the agent:
OTEL_EXPORTER_PROMETHEUS_PORT: "8888"
OTEL_METRICS_EXPORTER: prometheus

・Manual implementation:
java
SdkMeterProvider meterProvider = SdkMeterProvider.builder()
    .setResource(resource)
    .registerMetricReader(PrometheusHttpServer.builder().setPort(7777))
    .build();

**Reply by laurit** [reply]
If you do not wish to set up 2 prometheus servers then just don't. For the manual instrumentation use the opentelemetry instance from the agent that you can get by calling `GlobalOpenTelemetry.get()`.

**Reply by shinozakisatoshi11** [reply]
Thank you very much.

I wrote it as follows and created metrics using the meter I made, but it wasn't exported. (I have omitted some parts, but there are no errors.) Am I not using the agent settings correctly? Is there something more I need to do?

ーーーーjavaーーーー
openTelemetry = OpenTelemetryConfiguration.initOpenTelemetry();
meter = openTelemetry.getMeter("test");

LongCounter cardCounter = meter.counterBuilder("server.card.count")
    .setDescription("Counts the card API calls")
    .setUnit("1")
    .build();

cardCounter.add(1, Attributes.builder().put("Id", "Id").put("status", "SUCCESS").build());
